### ISSUE TITLE
Poor type inference for generic method type parameters in conditional expression

### ISSUE THREAD

**Issue by Gawen-pjr** [typetools/checker-framework#557]
Consider following code:

```java
Optional<String> opt()
{
    boolean flag = true;
    return flag ? Optional.of("Hello") : Optional.empty();
}
```

Checker framework raises an error:

> `incompatible types in return.`
> `found: @Initialized @NonNull Optional<@Initialized @Nullable String>`
> `required: @Initialized @NonNull Optional<@Initialized @NonNull String>`

If you add an explicit type parameter, no error is raised:

```java
Optional<String> opt()
{
   boolean flag = true;
   return flag ? Optional.of("Hello") : Optional.<String> empty();
}
```

But in Java 8, you are not required to specify type parameter here because type inference has been improved for complex expressions.

Checker Framework should be as smart as Java compiler on such expressions.

**Comment by wmdietl** [comment]
Thanks for reporting the false positive!
I'm able to reproduce this issue and added a test case at:
checker/tests/nullness/java8/Issue557.java

Method type argument inference works correctly when the method is on the top level. It seems to fail when it is part of a ternary expression.
We could further expand the test case to also test other new Java 8 locations, e.g. method type argument inference for a method call that is the argument to another generic method call.
